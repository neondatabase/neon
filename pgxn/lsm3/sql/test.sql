create extension lsm3;

create table t(k bigint, val bigint);
create index lsm3_index on t using lsm3(k);

set enable_seqscan=off;

insert into t values (1,10);
select lsm3_start_merge('lsm3_index');
select lsm3_wait_merge_completion('lsm3_index');
insert into t values (2,20);
select lsm3_start_merge('lsm3_index');
select lsm3_wait_merge_completion('lsm3_index');
insert into t values (3,30);
select lsm3_start_merge('lsm3_index');
select lsm3_wait_merge_completion('lsm3_index');
insert into t values (4,40);
select lsm3_start_merge('lsm3_index');
select lsm3_wait_merge_completion('lsm3_index');
insert into t values (5,50);
select lsm3_start_merge('lsm3_index');
select lsm3_wait_merge_completion('lsm3_index');
select lsm3_get_merge_count('lsm3_index');
select * from t where k = 1;
select * from t order by k;
select * from t order by k desc;
analyze t;
explain (COSTS OFF, TIMING OFF, SUMMARY OFF) select * from t order by k;

insert into t values (generate_series(1,100000), 1);
insert into t values (generate_series(1000001,200000), 2);
insert into t values (generate_series(2000001,300000), 3);
insert into t values (generate_series(1,100000), 1);
insert into t values (generate_series(1000001,200000), 2);
insert into t values (generate_series(2000001,300000), 3);
select * from t where k = 1;
select * from t where k = 1000000;
select * from t where k = 2000000;
select * from t where k = 3000000;
analyze t;
explain (COSTS OFF, TIMING OFF, SUMMARY OFF) select * from t where k = 1;
select lsm3_get_merge_count('lsm3_index') > 5;

truncate table t;
insert into t values (generate_series(1,1000000), 1);
select * from t where k = 1;

reindex table t;
select * from t where k = 1;

drop table t;

create table lsm(k bigint);
insert into lsm values (generate_series(1, 1000000));
create index concurrently on lsm using lsm3(k);
select * from lsm where k = 1;

drop table lsm;

