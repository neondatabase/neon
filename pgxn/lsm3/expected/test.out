create extension lsm3;
create table t(k bigint, val bigint);
create index lsm3_index on t using lsm3(k);
set enable_seqscan=off;
insert into t values (1,10);
select lsm3_start_merge('lsm3_index');
 lsm3_start_merge 
------------------
 
(1 row)

select lsm3_wait_merge_completion('lsm3_index');
 lsm3_wait_merge_completion 
----------------------------
 
(1 row)

insert into t values (2,20);
select lsm3_start_merge('lsm3_index');
 lsm3_start_merge 
------------------
 
(1 row)

select lsm3_wait_merge_completion('lsm3_index');
 lsm3_wait_merge_completion 
----------------------------
 
(1 row)

insert into t values (3,30);
select lsm3_start_merge('lsm3_index');
 lsm3_start_merge 
------------------
 
(1 row)

select lsm3_wait_merge_completion('lsm3_index');
 lsm3_wait_merge_completion 
----------------------------
 
(1 row)

insert into t values (4,40);
select lsm3_start_merge('lsm3_index');
 lsm3_start_merge 
------------------
 
(1 row)

select lsm3_wait_merge_completion('lsm3_index');
 lsm3_wait_merge_completion 
----------------------------
 
(1 row)

insert into t values (5,50);
select lsm3_start_merge('lsm3_index');
 lsm3_start_merge 
------------------
 
(1 row)

select lsm3_wait_merge_completion('lsm3_index');
 lsm3_wait_merge_completion 
----------------------------
 
(1 row)

select lsm3_get_merge_count('lsm3_index');
 lsm3_get_merge_count 
----------------------
                    5
(1 row)

select * from t where k = 1;
 k | val 
---+-----
 1 |  10
(1 row)

select * from t order by k;
 k | val 
---+-----
 1 |  10
 2 |  20
 3 |  30
 4 |  40
 5 |  50
(5 rows)

select * from t order by k desc;
 k | val 
---+-----
 5 |  50
 4 |  40
 3 |  30
 2 |  20
 1 |  10
(5 rows)

analyze t;
explain (COSTS OFF, TIMING OFF, SUMMARY OFF) select * from t order by k;
            QUERY PLAN            
----------------------------------
 Index Scan using lsm3_index on t
(1 row)

insert into t values (generate_series(1,100000), 1);
insert into t values (generate_series(1000001,200000), 2);
insert into t values (generate_series(2000001,300000), 3);
insert into t values (generate_series(1,100000), 1);
insert into t values (generate_series(1000001,200000), 2);
insert into t values (generate_series(2000001,300000), 3);
select * from t where k = 1;
 k | val 
---+-----
 1 |  10
 1 |   1
 1 |   1
(3 rows)

select * from t where k = 1000000;
 k | val 
---+-----
(0 rows)

select * from t where k = 2000000;
 k | val 
---+-----
(0 rows)

select * from t where k = 3000000;
 k | val 
---+-----
(0 rows)

analyze t;
explain (COSTS OFF, TIMING OFF, SUMMARY OFF) select * from t where k = 1;
            QUERY PLAN            
----------------------------------
 Index Scan using lsm3_index on t
   Index Cond: (k = 1)
(2 rows)

select lsm3_get_merge_count('lsm3_index') > 5;
 ?column? 
----------
 t
(1 row)

truncate table t;
insert into t values (generate_series(1,1000000), 1);
select * from t where k = 1;
 k | val 
---+-----
 1 |   1
(1 row)

reindex table t;
select * from t where k = 1;
 k | val 
---+-----
 1 |   1
(1 row)

drop table t;
create table lsm(k bigint);
insert into lsm values (generate_series(1, 1000000));
create index concurrently on lsm using lsm3(k);
select * from lsm where k = 1;
 k 
---
 1
(1 row)

drop table lsm;
